radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/service/reducer/AllValidators.java

```diff
@@ -0,0 +1,265 @@
+ */
+public final class AllValidators {
+	private final Map<ECPublicKey, ValidatorMetaData> metadata;
+	private final Map<ECPublicKey, UInt256> stake;
+	private final Map<ECPublicKey, REAddr> owners;
+	private final Map<ECPublicKey, Boolean> delegationFlags;
+	private final Map<ECPublicKey, Integer> rakes;
+	private final Set<ECPublicKey> registered;
+
+	private AllValidators(
+		Set<ECPublicKey> registered,
+		Map<ECPublicKey, UInt256> stake,
+		Map<ECPublicKey, REAddr> owners,
+		Map<ECPublicKey, Boolean> delegationFlags,
+		Map<ECPublicKey, ValidatorMetaData> metadata,
+		Map<ECPublicKey, Integer> rakes
+	) {
+		this.registered = registered;
+		this.stake = stake;
+		this.owners = owners;
+		this.delegationFlags = delegationFlags;
+		this.metadata = metadata;
+		this.rakes = rakes;
+	}
+
+	public static AllValidators create() {
+		return new AllValidators(Set.of(), Map.of(), Map.of(), Map.of(), Map.of(), Map.of());
+	}
+
+	public AllValidators set(ValidatorMetaData metaData) {
+		var nextMetadata = replaceEntry(Maps.immutableEntry(metaData.getValidatorKey(), metaData), metadata);
+		return new AllValidators(registered, stake, owners, delegationFlags, nextMetadata, rakes);
+	}
+
+	public AllValidators setRegistered(ECPublicKey validatorKey, boolean isRegistered) {
+		if (registered.contains(validatorKey) == isRegistered) {
+			return this;
+		}
+		if (!isRegistered) {
+			return new AllValidators(
+				registered.stream()
+					.filter(e -> !e.equals(validatorKey))
+					.collect(Collectors.toSet()),
+				stake,
+				owners,
+				delegationFlags,
+				metadata,
+				rakes
+			);
+		} else {
+			var set = ImmutableSet.<ECPublicKey>builder()
+				.addAll(registered)
+				.add(validatorKey)
+				.build();
+
+			return new AllValidators(set, stake, owners, delegationFlags, metadata, rakes);
+		}
+	}
+
+	public AllValidators add(ECPublicKey validatorKey) {
+		var set = ImmutableSet.<ECPublicKey>builder()
+			.addAll(registered)
+			.add(validatorKey)
+			.build();
+
+		return new AllValidators(set, stake, owners, delegationFlags, metadata, rakes);
+	}
+
+	public AllValidators remove(ECPublicKey validatorKey) {
+		return new AllValidators(
+			registered.stream()
+				.filter(e -> !e.equals(validatorKey))
+				.collect(Collectors.toSet()),
+			stake,
+			owners,
+			delegationFlags,
+			metadata,
+			rakes
+		);
+	}
+
+	public ValidatorMetaData getMetadata(ECPublicKey validatorKey) {
+		return ofNullable(metadata.get(validatorKey)).orElse(new ValidatorMetaData(validatorKey));
+	}
+
+	public REAddr getOwner(ECPublicKey validatorKey) {
+		return ofNullable(owners.get(validatorKey))
+			.orElse(REAddr.ofPubKeyAccount(validatorKey));
+	}
+
+	public UInt256 getStake(ECPublicKey validatorKey) {
+		return ofNullable(stake.get(validatorKey))
+			.orElse(UInt256.ZERO);
+	}
+
+	public Boolean allowsDelegation(ECPublicKey validatorKey) {
+		return ofNullable(delegationFlags.get(validatorKey))
+			.orElse(Boolean.TRUE);
```

Comment:
I was wondering if it would make more sense to return an `Optional` instead provides a default. Also I thought we changed the delegation flag default to `false`, didn't we?

Reply:
Yup, it should be defaulted to false

Reply:
Fixed